sdks/python/apache_beam/utils/counters.py

```diff
@@ -27,6 +27,46 @@
 from apache_beam.transforms import cy_combiners
 
 
+class CounterName(object):
+  """Naming information for a counter."""
+  SYSTEM = object()
+  USER = object()
+
+  def __init__(self, name, stage_name=None, step_name=None,
+               system_name=None, namespace=None,
+               origin=None, output_index=None):
+    self.name = name
+    self.origin = origin or CounterName.SYSTEM
+    self.namespace = namespace
+    self.stage_name = stage_name
+    self.step_name = step_name
+    self.system_name = system_name
+    self.output_index = output_index
+
+  def __hash__(self):
```

Comment:
The default hash function is related to the `id()` of the object. This means that without overriding it, this would be True: 
`hash(CounterName('arg1', 'arg2', 'arg3', 'arg4')) != hash(CounterName('arg1', 'arg2', 'arg3', 'arg4'))`

I need counters to be indexable by their name. e.g.
```
my_d = {CounterName('arg1', 'arg2', 'arg3', 'arg4'): my_counter}
assert my_d[CounterName('arg1', 'arg2', 'arg3', 'arg4')] == my_counter
```

That's why I need to override the hash function. In fact, I'll add unit tests for this.